src/topology/homotopy/homotopy_group.lean

```diff
@@ -5,141 +5,307 @@ Authors: Roberto Alvarez
 
-We show that `π 0 x` is equivalent to the path-conected components, and
-that `π 1 x` is equivalent to the fundamental group at `x`.
+We show that `π_0 x` is equivalent to the path-conected components, and
+that `π_1 x` is equivalent to the fundamental group at `x`.
+We give a group instance using path composition and show commutativity when `n > 1`.
 
 ## definitions
 
-* `gen_loop n x` is the type of continous fuctions `I^n → X` that send the boundary to `x`
-* `homotopy_group n x` denoted `π n x` is the quotient of `gen_loop n x` by homotopy relative
-  to the boundary
+* `gen_loop n x` is the type of continous fuctions `I^n → X` that send the boundary to `x`,
+* `homotopy_group n x` denoted `π_ n x` is the quotient of `gen_loop n x` by homotopy relative
+  to the boundary,
+* group instance `group (π_(n+1) x)`,
+* commutative group instance `comm_group (π_(n+2) x)`.
 
-TODO: show that `π n x` is a group when `n > 0` and abelian when `n > 1`. Show that
-`pi1_equiv_fundamental_group` is an isomorphism of groups.
+TODO: Path-induced homomorphisms. Show that `pi1_equiv_fundamental_group` is a group isomorphism.
+  Examples with 𝕊^n (π_n (𝕊^n) = ℤ, π_m (𝕊^n) trivial for m < n).
+  Actions of π_1 on π_n.
+  Group (up to homotopy) of Ω.
+  Lie algebra: ⁅π_(n+1), π_(m+1)⁆ contained in π_(n+m+1).
 
 -/
 
 open_locale unit_interval topology
+open homeomorph
 
 noncomputable theory
 
 universes u
 variables {X : Type u} [topological_space X]
-variables {n : ℕ} {x : X}
+variables {N : Type*} {x : X}
 
-/--
-The `n`-dimensional cube.
--/
+/-- We allow an arbitrary indexing type `N`, not just `fin n`, in the definition of `cube`, so as to
+  be able to identify facets of a cube to cubes indexed by subtypes of `N`, which is more
+  convenient than using `fin (n-1)`. -/
 @[derive [has_zero, has_one, topological_space]]
-def cube (n : ℕ) : Type := fin n → I
-local notation `I^` := cube
+def cube (N : Type*) := N → I
+local notation `I^` n := cube (fin n)
 
 namespace cube
 
-@[continuity] lemma proj_continuous (i : fin n) : continuous (λ f : I^n, f i) :=
-continuous_apply i
-
-/--
-The points of the `n`-dimensional cube with at least one projection equal to 0 or 1.
--/
-def boundary (n : ℕ) : set (I^n) := {y | ∃ i, y i = 0 ∨ y i = 1}
+instance compact_space : compact_space (cube N) :=
+by { convert pi.compact_space, intro, apply_instance }
 
-/--
-The first projection of a positive-dimensional cube.
--/
-@[simp] def head {n} : I^(n+1) → I := λ c, c 0
+instance locally_compact_space : locally_compact_space (cube N) :=
+by convert locally_compact_space.pi; intro; apply_instance
 
-@[continuity] lemma head.continuous {n} : continuous (@head n) := proj_continuous 0
+/-- The points in a cube with at least one projection equal to 0 or 1. -/
+def boundary (N) : set (cube N) := {y | ∃ i, y i = 0 ∨ y i = 1}
 
-/--
-The projection to the last `n` coordinates from an `n+1` dimensional cube.
--/
-@[simp] def tail {n} : I^(n+1) → I^n := λ c, fin.tail c
```

Comment:
Is there a reason you removed this?

Reply:
It wasn't being used anywhere, and it's just a synonym for `fin.tail`